trinity/sync/full/hexary_trie.py

```diff
@@ -0,0 +1,232 @@
+        without decoding and scheduling requests for children. This is needed to fetch contract
+        code when doing a state sync.
+        """
+        self.node_key = node_key
+        self.parents: List[SyncRequest] = []
+        if parent is not None:
+            self.parents = [parent]
+        self.depth = depth
+        self.leaf_callback = leaf_callback
+        self.is_raw = is_raw
+        self.dependencies = 0
+        self.data: bytes = None
+
+    def __lt__(self, other: 'SyncRequest') -> bool:
+        return self.depth < other.depth
+
+    def __repr__(self) -> str:
+        return "SyncRequest(%s, depth=%d)" % (encode_hex(self.node_key), self.depth)
+
+
+def _get_children(node: Hash32, depth: int
+                  ) -> Tuple[List[Tuple[int, Hash32]], List[bytes]]:
+    """Return all children of the node with the given hash.
+
+    :rtype: A two-tuple with one list containing the children that reference other nodes and
+    another containing the leaf children.
+    """
+    node_type = get_node_type(node)
+    references = []
+    leaves = []
+
+    if node_type == NODE_TYPE_BLANK:
+        pass
+    elif node_type == NODE_TYPE_LEAF:
+        leaves.append(node[1])
+    elif node_type == NODE_TYPE_EXTENSION:
+        if isinstance(node[1], bytes) and len(node[1]) == 32:
+            references.append((depth + 1, node[1]))
+        elif isinstance(node[1], list):
+            # the rlp encoding of the node is < 32 so rather than a 32-byte
+            # reference, the actual rlp encoding of the node is inlined.
+            sub_references, sub_leaves = _get_children(node[1], depth + 1)
+            references.extend(sub_references)
+            leaves.extend(sub_leaves)
+        else:
+            raise Exception("Invariant")
+    elif node_type == NODE_TYPE_BRANCH:
+        for sub_node in node[:16]:
+            if isinstance(sub_node, bytes) and len(sub_node) == 32:
+                # this is a reference to another node.
+                references.append((depth + 1, sub_node))
+            else:
+                sub_references, sub_leaves = _get_children(sub_node, depth)
+                references.extend(sub_references)
+                leaves.extend(sub_leaves)
+
+        # The last item in a branch may contain a value.
+        if not is_blank_node(node[16]):
+            leaves.append(node[16])
+
+    return references, leaves
+
+
+class HexaryTrieSync:
+
+    def __init__(self, root_hash: Hash32, db: AsyncBaseDB, logger: TraceLogger) -> None:
+        self.queue: List[SyncRequest] = []
+        self.requests: Dict[Hash32, SyncRequest] = {}
+        self.db = db
+        self.root_hash = root_hash
+        self.logger = logger
+        # A cache of node hashes we know to exist in our DB, used to avoid querying the DB
+        # unnecessarily as that's the main bottleneck when dealing with a large DB like for
+        # ethereum's mainnet/ropsten.
+        self._existing_nodes: Set[Hash32] = set()
```

Comment:
This worries me a bit as it has `O(n)` storage cost.  For a full state trie sync I believe that means storing roughly 180 million hashes which is 5gb of memory.  Do you think we can get by with an LRU cache or something on the existence check and not actually store all of them?

Reply:
I had a deja-vu feeling here so did a bit of digging and found https://github.com/ethereum/py-trie/pull/43#pullrequestreview-109756575

I've opened #1154 for it